src/vcpkg/base/system.mac.cpp

```diff
@@ -0,0 +1,262 @@
+                {
+                    return Hash::get_string_hash(mac, Hash::Algorithm::Sha256);
+                }
+            }
+        }
+#elif defined(__linux__) || defined(__APPLE__)
+        // The getifaddrs(ifaddrs** ifap) function creates a linked list of structures
+        // describing the network interfaces of the local system, and stores
+        // the address of the first item of the list in *ifap.
+        // man page: https://www.man7.org/linux/man-pages/man3/getifaddrs.3.html
+        struct ifaddrs_guard
+        {
+            ifaddrs* ptr = nullptr;
+            ~ifaddrs_guard()
+            {
+                if (ptr) freeifaddrs(ptr);
+            }
+        } interfaces;
+
+        if (getifaddrs(&interfaces.ptr) < 0)
+        {
+            return "0";
+        }
+
+        unsigned char bytes[MAC_BYTES_LENGTH];
+        for (auto interface = interfaces.ptr; interface; interface = interface->ifa_next)
+        {
+            // The ifa_addr field points to a structure containing the interface
+            // address.  (The sa_family subfield should be consulted to
+            // determine the format of the address structure.)  This field may
+            // contain a null pointer.
+            if (interface->ifa_addr && interface->ifa_addr->sa_family == AF_TYPE)
+            {
+                auto name = std::string(interface->ifa_name, strlen(interface->ifa_name));
+                if (interface->ifa_flags & IFF_LOOPBACK) continue;
+
+                // Convert the generic sockaddr into a specified representation
+                // based on the value of sa_family, on macOS the AF_PACKET
+                // family is not available so we fall back to AF_LINK.
+                // AF_PACKET and sockaddr_ll: https://man7.org/linux/man-pages/man7/packet.7.html
+                // AF_LINK and sockaddr_dl: https://illumos.org/man/3SOCKET/sockaddr_dl
+                std::memset(bytes, 0, MAC_BYTES_LENGTH);
+#if defined(__linux__)
+                auto address = reinterpret_cast<sockaddr_ll*>(interface->ifa_addr);
+                if (address->sll_halen != MAC_BYTES_LENGTH) continue;
+                std::memcpy(bytes, address->sll_addr, MAC_BYTES_LENGTH);
+#elif defined(__APPLE__)
+                auto address = reinterpret_cast<sockaddr_dl*>(interface->ifa_addr);
+                if (address->sdl_alen != MAC_BYTES_LENGTH) continue;
+                // The macro LLADDR() returns the start of the link-layer network address.
+                std::memcpy(bytes, LLADDR(address), MAC_BYTES_LENGTH);
+#endif
+                auto maybe_mac = mac_bytes_to_string(Span<unsigned char>(bytes, MAC_BYTES_LENGTH));
+                if (is_valid_mac_for_telemetry(maybe_mac))
+                {
+                    return Hash::get_string_hash(maybe_mac, Hash::Algorithm::Sha256);
+                }
+            }
+        }
+#else
+        // fallback when getifaddrs() is not available
+        struct socket_guard
```

Comment:
This is substantially identical to `PosixFd` in `files.cpp`, should they be merged?

Reply:
Didn't know we had that, I'll take a look

Reply:
Just pinging again on whether these can be merged.

Reply:
They can be merged but I don't feel like the effort to extract that struct for a simple use case like this one is worth it.